chainer/sequential.py

```diff
@@ -0,0 +1,383 @@
+
+    def __len__(self):
+        return len(self._layers)
+
+    def __getitem__(self, i):
+        return self._layers[i]
+
+    def __setitem__(self, i, layer):
+        if i >= len(self):
+            raise ValueError(
+                '{} should be less than {}'.format(i, len(self)))
+        if not callable(layer):
+            raise ValueError(
+                'All elements of a Sequential class should be callable. But '
+                'given {} is not callable.'.format(layer))
+
+        if self._layers[i] is not layer:
+            del self[i]
+            self.insert(i, layer)
+
+    def __delitem__(self, i):
+        layer = self._layers.pop(i)
+        if isinstance(layer, chainer.Link):
+            for i, link in enumerate(self._children):
+                if link is layer:
+                    del self._children[i]
+                    break
+        elif callable(layer) and hasattr(layer, '__name__') \
+                and layer.__name__ == '<lambda>':
+            self._n_lambda -= 1
+
+    def __iter__(self):
+        return iter(self._layers)
+
+    def __reversed__(self):
+        return reversed(self._layers)
+
+    def __contains__(self, item):
+        return item in self._layers
+
+    def __add__(self, other):
+        if isinstance(other, Sequential):
+            ret = Sequential()
+            for layer in self:
+                ret.append(layer)
+            for layer in other:
+                ret.append(layer)
+            return ret
+        elif callable(other):
+            self.append(other)
+            return self
+        else:
+            raise ValueError('add (+) operator is support only callable '
+                             'objects, but {} is not callable'.format(
+                                 str(type(other))))
+
+    def __radd__(self, other):
+        if callable(other):
+            return Sequential(other) + self
+        else:
+            raise ValueError('add (+) operator is support only callable '
+                             'objects, but {} is not callable'.format(
+                                 str(type(other))))
+
+    def __iadd__(self, other):
+        if isinstance(other, Sequential):
+            for layer in other:
+                self.append(layer)
+        elif callable(other):
+            self.append(other)
+        else:
+            raise ValueError('add (+) operator is support only callable '
+                             'objects, but {} is not callable'.format(
+                                 str(type(other))))
+        return self
+
+    def __mul__(self, n_repeat):
+        if n_repeat <= 0:
+            return Sequential()
+        ret = self.copy()
+        for _ in range(n_repeat - 1):
+            for layer in self:
+                if isinstance(layer, chainer.Link):
+                    layer = copy.deepcopy(layer)
+                    for param in layer.params(include_uninit=False):
+                        param.initialize(param.shape)
```

Comment:
Is it safe to assume that `param` has an initializer?

Reply:
By the change in [chainer/variable.py L:1203](https://github.com/chainer/chainer/pull/2918/files/f6ba4799086872f636dbabc9e8de1c73ca897795#diff-e131a8dfc2ea641e7e7c63597b9b0d8eR1203), `param` which is a `Parameter` object always has initializer.

Reply:
If a user sets a random array as the initial value, it's detected as a `Constant` initializer and then the parameters will not be sampled again when multiplying the Sequential containing that link. Is it appropriate? One can avoid this kind of thing by enforcing some coding practice (e.g. always use an appropriate initializer (or define one) when you initialize a parameter randomly), but I'm not sure if it's common to users (at least we did not advertise this coding practice).

Reply:
Also, as I wrote in the above comment of the PR thread, this behavior (re-initializing the parameters on multiplying) is not consistent with the behavior of `list`, which does not copy the content. In Python, we usually use list comprehension in such a case to repeat the same but independently initialized values in a list, like `[Linear(100) for _ in range(10)]` instead of `[Linear(100)] * 10`. I understand that the original behavior of `list.__mul__` is confusing (e.g. 10 Linears with _shared_ weights in this example), so it's ok to not reproduce it in `Sequential`, but I'm wondering if it's appropriate to provide an alternative definition of `__mul__`.

Reply:
Do we need to ask other members as well? My opinion may be extreme...